File: src/main/java/com/company/erp/features/users/dto/UserDto.java
package com.company.erp.features.users.dto;

import com.company.erp.features.users.entity.Role;
import com.company.erp.features.users.entity.User;

import java.util.Set;
import java.util.stream.Collectors;

public record UserDto(
        Long id,
        String email,
        String fullName,
        String phone,
        String avatarUrl,
        boolean enabled,
        Set<String> roles,
        Set<String> permissions
) {
    public static UserDto from(User u) {
        return new UserDto(
                u.getId(),
                u.getEmail(),
                u.getFullName(),
                u.getPhone(),
                u.getAvatarUrl(),
                u.isEnabled(),
                u.getRoles().stream().map(Role::getCode).collect(Collectors.toSet()),
                u.allPermissions());
    }
}

File: app/Features/Users/Dto/UserDto.php
<?php

namespace App\Features\Users\Dto;

use App\Features\Users\Models\User;

/** Port of the Spring UserDto record. */
final class UserDto
{
    /**
     * @return array<string, mixed>
     */
    public static function from(User $u): array
    {
        return [
            'id' => (int) $u->id,
            'email' => $u->email,
            'fullName' => $u->full_name,
            'phone' => $u->phone,
            'avatarUrl' => $u->avatar_url,
            'enabled' => (bool) $u->enabled,
            'roles' => array_values($u->roleCodes()),
            'permissions' => array_values($u->allPermissions()),
        ];
    }
}
